httpserver/server.go

```diff
@@ -482,6 +482,27 @@ func (s *Server) createAuthenticatedContext(r *http.Request) (context.Context, e
 	}
 	ctx = htcondor.WithSecurityConfig(ctx, secConfig)
 
+	// Extract username for rate limiting
+	var username string
+	if s.userHeader != "" {
+		// Check if using user header mode
+		_, bearerErr := extractBearerToken(r)
+		if bearerErr != nil {
+			// User header mode - get username from header
+			username = r.Header.Get(s.userHeader)
+		}
+	}
+
+	// If username not from header, try to extract from JWT token
+	if username == "" {
+		username, _ = parseJWTUsername(token)
```

Comment:
Tokens must be validated prior to being used for rate limiting.

Keep a cache of tokens that previously successfully authenticated (periodically remove expired tokens from cache). I think there's an example of this already in httpserver. Only use the authenticated user code if the token matches one in the cache.

Otherwise, treat this as an unauthenticated user.

Reply:
(Refer to the logic in the MCP server)